db/nft.go

```diff
@@ -183,3 +235,97 @@ func GetNftEvents(conn *pgxpool.Conn, q QueryEventsRequest, p PageRequest) (Quer
 	res.Pagination.Count = len(res.Events)
 	return res, nil
 }
+
+func GetCollector(conn *pgxpool.Conn, q QueryCollectorRequest) (res QueryCollectorResponse, err error) {
+	sql := `
+	SELECT owner, sum(count) as total,
+		array_agg(json_build_object(
+			'iscn_id_prefix', iscn_id_prefix,
+			'class_id', class_id,
+			'count', count))
+	FROM (
+		SELECT n.owner, i.iscn_id_prefix, c.class_id, COUNT(DISTINCT n.id) as count
+		FROM iscn as i
+		JOIN nft_class as c ON i.iscn_id_prefix = c.parent_iscn_id_prefix
+		JOIN nft AS n ON c.class_id = n.class_id
+		WHERE i.owner = $1
+		GROUP BY n.owner, i.iscn_id_prefix, c.class_id
+	) as r
+	GROUP BY owner
+	ORDER BY total DESC
+	OFFSET $2
+	LIMIT $3
```

Comment:
Need to add indexes on used fields (e.g. `iscn.iscn_id_prefix`, `nft.class_id`, `iscn.data #>> '{"contentMetadata", "@type"}'`, `nft_class.created_at`), otherwise it's always doing full table scan on various tables.
Currently the `iscn` table is the largest, so especially need to avoid full table scan on `iscn` table.

You may check by running `EXPLAIN` and `EXPLAIN ANALYSE`. `Seq Scan on blablabla_table` means Postgres is doing full table scan, and `EXPLAIN ANALYSE` will give you the actual number of rows got, so if the number of rows goes > 10000 then you may need to check whether you should build index / optimize on the query.

Reply:
Currently if I query with no condition and LIMIT = 10, `EXPLAIN ANALYSE` gives the following output:
```
EXPLAIN ANALYSE
	SELECT c.class_id, c.name, c.description, c.symbol, c.uri, c.uri_hash,
	c.config, c.metadata, c.price,
	c.parent_type, c.parent_iscn_id_prefix, c.parent_account, c.created_at, sold_count
	FROM nft_class as c
	JOIN (
		SELECT c.id, count(n.id) as sold_count, array_remove(array_agg(DISTINCT n.owner), NULL) as owners
		FROM iscn as i
		JOIN nft_class as c ON i.iscn_id_prefix = c.parent_iscn_id_prefix
		LEFT JOIN nft as n ON c.class_id = n.class_id
		WHERE
      ('' = '' OR i.owner = '')
			AND (false = true OR n.owner != i.owner)
			AND (NULL::text[] IS NULL OR n.owner != ALL(NULL))
			AND ('' = '' OR i.data #>> '{"contentMetadata", "@type"}' = '')
			AND (NULL::jsonb IS NULL OR i.stakeholders @> NULL)
			AND (NULL::jsonb IS NULL OR i.stakeholders @> NULL)
			AND (0 = 0 OR c.created_at > to_timestamp(0))
			AND (0 = 0 OR c.created_at < to_timestamp(0))
		GROUP BY c.id
	) AS t USING(id)
	WHERE ('' = '' OR '' = ANY(t.owners))
	ORDER BY sold_count DESC
	LIMIT 10
;


Limit  (cost=455964.35..455964.38 rows=10 width=344) (actual time=1282.020..1282.023 rows=10 loops=1)
  ->  Sort  (cost=455964.35..455964.88 rows=210 width=344) (actual time=1253.995..1253.996 rows=10 loops=1)
        Sort Key: (count(n.id)) DESC
        Sort Method: top-N heapsort  Memory: 50kB
        ->  Hash Join  (cost=455844.35..455959.82 rows=210 width=344) (actual time=1250.053..1253.928 rows=34 loops=1)
              Hash Cond: (c_1.id = c.id)
              ->  Finalize GroupAggregate  (cost=455829.62..455942.43 rows=210 width=48) (actual time=1249.970..1253.809 rows=34 loops=1)
                    Group Key: c_1.id
                    ->  Gather Merge  (cost=455829.62..455939.28 rows=210 width=16) (actual time=1249.609..1260.460 rows=68 loops=1)
                          Workers Planned: 1
                          Workers Launched: 1
                          ->  Partial GroupAggregate  (cost=454829.61..454915.64 rows=210 width=16) (actual time=1212.827..1216.320 rows=34 loops=2)
                                Group Key: c_1.id
                                ->  Sort  (cost=454829.61..454857.59 rows=11191 width=16) (actual time=1212.457..1213.727 rows=17495 loops=2)
                                      Sort Key: c_1.id
                                      Sort Method: quicksort  Memory: 1592kB
                                      Worker 0:  Sort Method: quicksort  Memory: 1585kB
                                      ->  Parallel Hash Join  (cost=452079.17..454077.01 rows=11191 width=16) (actual time=1198.666..1208.786 rows=17495 loops=2)
                                            Hash Cond: (n.class_id = c_1.class_id)
                                            Join Filter: (n.owner <> i.owner)
                                            Rows Removed by Join Filter: 6
                                            ->  Parallel Seq Scan on nft n  (cost=0.00..1902.01 rows=10001 width=119) (actual time=0.024..2.169 rows=8500 loops=2)
                                            ->  Parallel Hash  (cost=452077.94..452077.94 rows=98 width=86) (actual time=1198.447..1198.447 rows=36 loops=2)
                                                  Buckets: 1024  Batches: 1  Memory Usage: 40kB
                                                  ->  Hash Join  (cost=14.72..452077.94 rows=98 width=86) (actual time=581.055..1197.376 rows=36 loops=2)
                                                        Hash Cond: (i.iscn_id_prefix = c_1.parent_iscn_id_prefix)
                                                        ->  Parallel Seq Scan on iscn i  (cost=0.00..425147.61 rows=936161 width=112) (actual time=0.169..1032.311 rows=1123075 loops=2)
                                                        ->  Hash  (cost=12.10..12.10 rows=210 width=72) (actual time=10.183..10.184 rows=37 loops=2)
                                                              Buckets: 1024  Batches: 1  Memory Usage: 15kB
                                                              ->  Seq Scan on nft_class c_1  (cost=0.00..12.10 rows=210 width=72) (actual time=10.153..10.170 rows=37 loops=2)
              ->  Hash  (cost=12.10..12.10 rows=210 width=344) (actual time=0.072..0.072 rows=37 loops=1)
                    Buckets: 1024  Batches: 1  Memory Usage: 52kB
                    ->  Seq Scan on nft_class c  (cost=0.00..12.10 rows=210 width=344) (actual time=0.013..0.050 rows=37 loops=1)
Planning Time: 0.872 ms
JIT:
  Functions: 68
  Options: Inlining false, Optimization false, Expressions true, Deforming true
  Timing: Generation 6.742 ms, Inlining 0.000 ms, Optimization 2.205 ms, Emission 45.504 ms, Total 54.450 ms
Execution Time: 1342.542 ms
```

You can see that there are full table scans (`Seq Scan`) on `nft_class`, `nft` and `iscn`.
Note that not every full table scan are bad, but we want to avoid them as much as possible, especially the one on `iscn` table.
(It is for the JOIN on `(i.iscn_id_prefix = c_1.parent_iscn_id_prefix)`, so an index on `iscn.iscn_id_prefix` should help)

Reply:
Things looked much better after building the index:

```
CREATE INDEX idx_iscn_iscn_id_prefix on iscn (iscn_id_prefix);

-- repeat the `EXPLAIN ANALYSE` above

Limit  (cost=4210.08..4210.11 rows=10 width=344) (actual time=63.031..63.034 rows=10 loops=1)
  ->  Sort  (cost=4210.08..4210.61 rows=210 width=344) (actual time=63.029..63.030 rows=10 loops=1)
        Sort Key: t.sold_count DESC
        Sort Method: top-N heapsort  Memory: 50kB
        ->  Hash Join  (cost=4192.88..4205.55 rows=210 width=344) (actual time=62.858..62.910 rows=34 loops=1)
              Hash Cond: (c.id = t.id)
              ->  Seq Scan on nft_class c  (cost=0.00..12.10 rows=210 width=344) (actual time=0.006..0.021 rows=37 loops=1)
              ->  Hash  (cost=4190.26..4190.26 rows=210 width=16) (actual time=62.830..62.830 rows=34 loops=1)
                    Buckets: 1024  Batches: 1  Memory Usage: 10kB
                    ->  Subquery Scan on t  (cost=4186.06..4190.26 rows=210 width=16) (actual time=62.785..62.806 rows=34 loops=1)
                          ->  HashAggregate  (cost=4186.06..4188.16 rows=210 width=48) (actual time=62.783..62.797 rows=34 loops=1)
                                Group Key: c_1.id
                                ->  Hash Join  (cost=1817.36..4090.94 rows=19024 width=16) (actual time=25.560..53.174 rows=34990 loops=1)
                                      Hash Cond: (n.class_id = c_1.class_id)
                                      Join Filter: (n.owner <> i.owner)
                                      Rows Removed by Join Filter: 11
                                      ->  Seq Scan on nft n  (cost=0.00..1972.01 rows=17001 width=119) (actual time=0.011..6.902 rows=17001 loops=1)
                                      ->  Hash  (cost=1814.42..1814.42 rows=235 width=86) (actual time=25.430..25.430 rows=73 loops=1)
                                            Buckets: 1024  Batches: 1  Memory Usage: 20kB
                                            ->  Nested Loop  (cost=0.56..1814.42 rows=235 width=86) (actual time=0.435..25.348 rows=73 loops=1)
                                                  ->  Seq Scan on nft_class c_1  (cost=0.00..12.10 rows=210 width=72) (actual time=0.002..0.036 rows=37 loops=1)
                                                  ->  Index Scan using idx_iscn_iscn_id_prefix on iscn i  (cost=0.56..8.57 rows=1 width=112) (actual time=0.679..0.681 rows=2 loops=37)
                                                        Index Cond: (iscn_id_prefix = c_1.parent_iscn_id_prefix)
Planning Time: 2.057 ms
Execution Time: 63.223 ms
```

But there are still full table scans on other tables, currently they are small so full table scan is still fast, may need to monitor and build index in the future.

Reply:
Should I based on #35 to make those index?

Reply:
I think it should be base on the new schema and migration structure, but #35 is still work in progress...